x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java

```diff
@@ -204,6 +215,92 @@ public static String exceptionToString(Exception ex) {
+        long numInProgress = 0;
+        for (final List<SnapshotsInProgress.Entry> entriesForRepo : snapshots.entriesByRepo()) {
+            numInProgress += entriesForRepo.stream().map(SnapshotsInProgress.Entry::userMetadata)
+                .filter(Objects::nonNull)
+                .map(meta -> meta.get(SnapshotsService.POLICY_ID_METADATA_FIELD))
+                .filter(policyId::equals)
+                .count();
+        }
+        return numInProgress;
+    }
+
+    private static class PreRegisterSLMRun extends ClusterStateUpdateTask {
+        private final String policyName;
+        private final String snapshotName;
+
+        PreRegisterSLMRun(String policyName, String snapshotName) {
+            this.policyName = policyName;
+            this.snapshotName = snapshotName;
+        }
+
+        @Override
+        public ClusterState execute(ClusterState currentState) throws Exception {
+            SnapshotLifecycleMetadata snapMeta = currentState.metadata().custom(SnapshotLifecycleMetadata.TYPE);
+
+            assert snapMeta != null : "this should never be called while the snapshot lifecycle cluster metadata is null";
+            if (snapMeta == null) {
+                logger.error(
+                    "failed to pre-register snapshot [{}] in policy [{}]: snapshot lifecycle metadata is null",
+                    snapshotName,
+                    policyName
+                );
+                return currentState;
+            }
+
+            Map<String, SnapshotLifecyclePolicyMetadata> snapLifecycles = new HashMap<>(snapMeta.getSnapshotConfigurations());
+            SnapshotLifecyclePolicyMetadata policyMetadata = snapLifecycles.get(policyName);
+            if (policyMetadata == null) {
+                logger.warn(
+                    "failed to pre-register snapshot [{}] in policy [{}]: policy not found",
+                    snapshotName,
+                    policyName
+                );
+                return currentState;
+            }
+
+            long preRegisteredRuns = policyMetadata.getPreRegisteredRuns();
+            SnapshotsInProgress snapshots = currentState.custom(SnapshotsInProgress.TYPE);
+            final long numSnapshotsRunning = snapshots == null ? 0 : numPolicySnapshotsInProgress(policyName, snapshots);
+
+            SnapshotLifecyclePolicyMetadata.Builder newPolicyMetadata = SnapshotLifecyclePolicyMetadata.builder(policyMetadata);
+            if (preRegisteredRuns > numSnapshotsRunning) {
+                final long unrecordedFailures = preRegisteredRuns - numSnapshotsRunning;
+                newPolicyMetadata.setInvocationsSinceLastSuccess(policyMetadata.getInvocationsSinceLastSuccess() + unrecordedFailures);
+
+                // There are likely scenarios where inc/decrements to preRegisteredRuns can be lost, so lower bound to 1 before run.
+                long newPreRegisteredRuns = Math.min(1, preRegisteredRuns + 1 - unrecordedFailures);
```

Comment:
It seem there are likely situations where `preRegisteredRuns` and `numSnapshotsRunning` are inconsistent and thus `unrecordedFailures` gets a wrong value. Setting  `newPreRegisteredRuns` to a minimum value of 1 should mitigate this, but doesn't solve it.

One alternative to improve this would be to store a set of pre-registered snapshot ids rather than a single count. Then remove all snapshot ids that are currently running from the set. Use the size of the remaining set to increment invocationsSinceLastSuccess.